File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
     List<List<Integer>> ans = new ArrayList<>();
    public void dfs(int[] candidates, int target, List<Integer> currSubarr, int start){
        if(target < 0){
            return;// If remain becomes negative, terminate the current branch of the backtracking
        }
        else if(target == 0){
            //Deep copy
            ans.add(new ArrayList<>(currSubarr));// If remain is zero, add the current combination to the result
        }
        else{
            int prevElement = Integer.MIN_VALUE;
            for(int i = start; i < candidates.length; i++){
                //if(i > start && candidates[i] == candidates[i-1]) continue;
                if(prevElement == candidates[i]) continue;
                if(candidates[i] <= target){
                    currSubarr.add(candidates[i]);// Choose the current candidate
                    dfs(candidates, target-candidates[i], currSubarr, i+1);// Recursively backtrack with updated remain
                    currSubarr.remove(currSubarr.size()-1);// Remove the last element to backtrack
                    prevElement = candidates[i];
                }
            }
        }
    }
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        if(target == 0){
            return ans;
        }
        List<Integer> curr = new ArrayList<>();
        Arrays.sort(candidates);
        dfs(candidates, target, curr, 0);

        return ans;
    }
}

File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    void backtrack(vector<int>& candidates, int target, int start, vector<int>& path, vector<vector<int>>& result) {
        // Base case: If the target is reached, add the current combination to the result
        if (target == 0) {
            result.push_back(path);
            return;
        }

        //int prevElm = INT_MIN; 

        // Iterate through the candidates starting from the 'start' index
        for (int i = start; i < candidates.size(); ++i) {
            // Skip duplicates to avoid duplicate combinations
            // This checks if the current element is the same as the previous one, and if it is, it skips it, but only when i > start.
            if (i > start && candidates[i] == candidates[i - 1]) {
                continue;
            }

            //if(candidates[i] == prevElm) continue;

            // Check if the current candidate can be included in the combination
            if (candidates[i] <= target) {
                // Add the current candidate to the combination
                path.push_back(candidates[i]);
                // Recursively explore further combinations with the remaining target
                backtrack(candidates, target - candidates[i], i + 1, path, result);
                // Backtrack: Remove the last added candidate to try other combinations
                path.pop_back();
            }

            //prevElm = candidates[i];
        }
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        ios_base::sync_with_stdio(false);  
        cin.tie(NULL);
        
        vector<vector<int>> result;
        vector<int> path;
        // Sort the candidates to handle duplicates efficiently
        sort(candidates.begin(), candidates.end());
        // Start backtracking from the beginning
        backtrack(candidates, target, 0, path, result);
        return result;
    }
};